### PR TITLE
oss-fuzz: build libcap and libmount along with systemd

### PR DESCRIPTION
**Pull request by evverx**
https://github.com/systemd/systemd/issues/20542

### CODE REVIEW

**Comment by yuwata** [comment]
Oh, nice. If this works as expected, then #25844 should not be necessary.

File: tools/oss-fuzz.sh [hunk 1]
```diff
@@ -71,6 +72,22 @@ else
...
+    popd
+
+    git clone https://github.com/util-linux/util-linux
+    pushd util-linux
+    git checkout 1f5129b79ad232c79ecbac31998e96c20ff4c90c # v2.38
+    ./autogen.sh
+    ./configure --prefix=/usr --disable-all-programs --enable-libuuid --enable-libblkid --enable-libmount
+    make -j"$(nproc)" V=1
+    make install V=1
+    popd
```
Comment (evverx) [inline tools/oss-fuzz.sh:92]: Note to self: to judge from
```
2022-12-23T07:48:47.3884826Z meson.build:1046:0: ERROR: C shared or static library 'crypt' not found
2022-12-23T07:48:48.2005404Z 2022-12-23 07:48:48,196 - root - ERROR - Building fuzzers failed.
2022-12-23T07:48:48.2005927Z 2022-12-23 07:48:48,197 - root - ERROR - Error building fuzzers for (commit: 43b7ed29275704e92ed004a2eeba2e4e5319d28e, pr_ref: refs/pull/25842/merge).
2022-12-23T07:48:48.4523533Z Cleaning up orphan processes
```
`libxcrypt` should also be built.

"i386" (which is actually "x86_64" with `-m32`) is a total mess at the moment.

**Comment by keszybz** [comment]
Looks reasonable, but let's wait for https://github.com/google/oss-fuzz/issues/9280 to be resolved.

**Comment by yuwata** [comment]
@evverx CIFuzz also failed with similar issue: https://github.com/systemd/systemd/actions/runs/3926572689/jobs/6712410038
```
==3871==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x7fa59f48f6f5 in update_argv /work/build/../../src/systemd/src/basic/argv-util.c:93:13
    #1 0x7fa59f48f6f5 in rename_process /work/build/../../src/systemd/src/basic/argv-util.c:200:16
    #2 0x7fa59f5ef59c in safe_fork_full /work/build/../../src/systemd/src/basic/process-util.c:1225:21
    #3 0x7fa59f67223b in safe_fork /work/build/../../src/systemd/src/basic/process-util.h:163:16
    #4 0x7fa59f67223b in parse_timestamp /work/build/../../src/systemd/src/basic/time-util.c:915:13
    #5 0x4a61c1 in LLVMFuzzerTestOneInput /work/build/../../src/systemd/src/fuzz/fuzz-time-util.c:17:16
    #6 0x4c49a3 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:611:15
    #7 0x4c418a in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool, bool*) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:514:3
    #8 0x4c4025 in fuzzer::Fuzzer::RereadOutputCorpus(unsigned long) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:431:11
    #9 0x4c646d in fuzzer::Fuzzer::Loop(std::__Fuzzer::vector<fuzzer::SizedFile, std::__Fuzzer::allocator<fuzzer::SizedFile> >&) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:873:7
    #10 0x4b588f in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:912:6
    #11 0x4deee2 in main /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerMain.cpp:20:10
    #12 0x7fa59ea40082 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24082) (BuildId: 1878e6b475720c7c51969e69ab2d276fae6d1dee)
    #13 0x41f57d in _start (build-out/fuzz-time-util+0x41f57d)

DEDUP_TOKEN: update_argv--rename_process--safe_fork_full
  Uninitialized value was created by an allocation of 'fv' in the stack frame of function 'have_effective_cap'
    #0 0x7fa59f494d00 in have_effective_cap /work/build/../../src/systemd/src/basic/capability-util.c:20

DEDUP_TOKEN: have_effective_cap
SUMMARY: MemorySanitizer: use-of-uninitialized-value /work/build/../../src/systemd/src/basic/argv-util.c:93:13 in update_argv
```

Does this PR also fix the issue? I am not familiar, but do both oss-fuzz and CIFuzz use the same script?

**Comment by evverx** [comment]
> Does this PR also fix the issue?

Yes, it should fix that as well. CIFuzz shouldn't have failed there though because that particular crash should be reproducible with the latest systemd build but as far as I can tell something went wrong when CIFuzz tried to reproduce it:
```
2023-01-16T02:51:28.6136157Z 2023-01-16 02:51:28,612 - root - INFO - Reproduce command returned: -9. Reproducible on /github/workspace/build-out/fuzz-time-util.
2023-01-16T02:51:28.6137166Z 2023-01-16 02:51:28,613 - root - INFO - Crash is reproducible.
2023-01-16T02:51:28.8413090Z 2023-01-16 02:51:28,840 - root - INFO - Downloading latest build.
2023-01-16T02:51:28.8445750Z 2023-01-16 02:51:28,843 - urllib3.connectionpool - DEBUG - Starting new HTTPS connection (1): storage.googleapis.com:443
2023-01-16T02:51:29.1377350Z 2023-01-16 02:51:29,135 - urllib3.connectionpool - DEBUG - https://storage.googleapis.com:443 "GET /clusterfuzz-builds/systemd/systemd-memory-202301160000.zip HTTP/1.1" 200 87974523
2023-01-16T02:51:32.0988322Z 2023-01-16 02:51:32,097 - root - INFO - Done downloading latest build.
2023-01-16T02:51:32.0996123Z 2023-01-16 02:51:32,099 - root - INFO - Trying to reproduce crash using: /tmp/tmp2wgtewve/crash-6d4fcd1b160983503df5535e6ca773a28a7672b4.
2023-01-16T02:52:02.1322664Z 2023-01-16 02:52:02,131 - root - ERROR - Reproducing timed out..
2023-01-16T02:52:02.1348623Z 2023-01-16 02:52:02,134 - root - INFO - The crash is not reproducible on previous build. Code change (pr/commit) introduced crash.
```
That issue was also found in https://github.com/systemd/systemd/actions/runs/3927106985/jobs/6713430013 but it wasn't reported there:
```
2023-01-16T04:38:15.9290930Z 2023-01-16 04:38:15,928 - root - ERROR - Reproducing timed out..
2023-01-16T04:38:15.9294270Z 2023-01-16 04:38:15,928 - root - INFO - Crash is not reproducible.
```
I suspect it has something to do with https://github.com/google/oss-fuzz/issues/9318.

**Comment by evverx** [comment]
Judging by https://github.com/google/oss-fuzz/pull/9428 it seems I haven't pushed the latest version of this PR where I fixed coverage builds yet.

In the meantime the fuzz targets keep falling apart: https://github.com/google/oss-fuzz/issues/9431. That is pretty bad given that libFuzzer is no longer actively maintained https://llvm.org/docs/LibFuzzer.html#status and OSS-Fuzz is slowly switching to centipede.

**Comment by yuwata** [comment]
> Looks reasonable, but let's wait for [google/oss-fuzz#9280](https://github.com/google/oss-fuzz/issues/9280) to be resolved.

> I suspect it has something to do with [google/oss-fuzz#9318](https://github.com/google/oss-fuzz/issues/9318).

@evverx Both two issues are resolved. Do you think this PR is ready now?

**Comment by evverx** [comment]
@yuwata I'd wait for https://github.com/google/oss-fuzz/pull/9427#issuecomment-1384195091 to get fixed. Right now that fuzzing engine isn't compatible with `meson` and it would be easier to make it work without this PR. Once that issue is resolved I'll try to make this PR work with `sanitizer=none` (which is passed to build unsanitized fuzz targets for centipede).

**Comment by evverx** [comment]
I put this on hold. I'll go ahead and close it.

In the meantime https://github.com/google/oss-fuzz/issues/9431 got fixed so systemd no longer fails to compile with centipede.

**Pull request closed without merging**